### ISSUE TITLE
Process instances can be created for deleted process

### ISSUE THREAD

**Issue by korthout** [camunda/camunda#21357]
**Describe the bug**

There is a race condition in multi-partition clusters between resource deployment and resource deletion. The condition is very specific and has a low likelihood but can be reproduced. After deleting a resource, the resource may become available again on non-deployment partitions.

Both resource deployment and resource deletion are distributed among partitions via command distribution. Command distribution does not guarantee the order of commands, so processors of distributed command must be idempotent.

If a deployment is distributed successfully to another partition, but this command distribution is not acknowledged, the command is redistributed. If the resource is also deleted, the deployment may be created again after it was deleted.

<details><summary>Compact record log</summary>
<p>

After the resource is deleted from all partitions, the process is recreated on partitions 2 and 3 because the `DEPLOYMENT CREATE` command distribution was not acknowledged.

```
1 C DPLY              CREATE       - #01-> -1    -1 - 
1 E PROC              CREATED      - #02->#01 P1K02 - process.xml -> "id-e756..45a8a2c" (version:1)
1 E DPLY              CREATED      - #03->#01 P1K01 - process.xml
1 E DSTR              STARTED      - #04->#01 P1K01 - DEPLOYMENT CREATE on partition 1
1 E DSTR              DISTRIBUTING - #05->#01 P1K01 - DEPLOYMENT CREATE to partition 2
1 E DSTR              DISTRIBUTING - #06->#01 P1K01 - DEPLOYMENT CREATE to partition 3
2 C DPLY              CREATE       - #01-> -1 P1K01 - process.xml
3 C DPLY              CREATE       - #01-> -1 P1K01 - process.xml
2 E PROC              CREATED      - #02->#01 P1K02 - process.xml -> "id-e756..45a8a2c" (version:1)
2 E DPLY              CREATED      - #03->#01 P1K01 - process.xml
3 E PROC              CREATED      - #02->#01 P1K02 - process.xml -> "id-e756..45a8a2c" (version:1)
3 E DPLY              CREATED      - #03->#01 P1K01 - process.xml
1 C RESOURCE_DELETION DELETE       - #07-> -1    -1 - ResourceDeletionRecord {"tenantId":"","resourceKey":2251799813685250}
1 E RESOURCE_DELETION DELETING     - #08->#07 P1K03 - ResourceDeletionRecord {"tenantId":"","resourceKey":2251799813685250}
1 E PROC              DELETING     - #09->#07 P1K04 - process.xml -> "id-e756..45a8a2c" (version:1)
1 E PROC              DELETED      - #10->#07 P1K05 - process.xml -> "id-e756..45a8a2c" (version:1)
1 E RESOURCE_DELETION DELETED      - #11->#07 P1K03 - ResourceDeletionRecord {"tenantId":"<default>","resourceKey":2251799813685250}
1 E DSTR              STARTED      - #12->#07 P1K03 - RESOURCE_DELETION DELETE on partition 1
1 E DSTR              DISTRIBUTING - #13->#07 P1K03 - RESOURCE_DELETION DELETE to partition 2
1 E DSTR              DISTRIBUTING - #14->#07 P1K03 - RESOURCE_DELETION DELETE to partition 3
2 C RESOURCE_DELETION DELETE       - #04-> -1 P1K03 - ResourceDeletionRecord {"tenantId":"<default>","resourceKey":2251799813685250}
3 C RESOURCE_DELETION DELETE       - #04-> -1 P1K03 - ResourceDeletionRecord {"tenantId":"<default>","resourceKey":2251799813685250}
2 E RESOURCE_DELETION DELETING     - #05->#04 P1K03 - ResourceDeletionRecord {"tenantId":"<default>","resourceKey":2251799813685250}
2 E PROC              DELETING     - #06->#04 P2K01 - process.xml -> "id-e756..45a8a2c" (version:1)
2 E PROC              DELETED      - #07->#04 P2K02 - process.xml -> "id-e756..45a8a2c" (version:1)
2 E RESOURCE_DELETION DELETED      - #08->#04 P1K03 - ResourceDeletionRecord {"tenantId":"<default>","resourceKey":2251799813685250}
1 C DSTR              ACKNOWLEDGE  - #15-> -1 P1K03 - RESOURCE_DELETION DELETE for partition 2
1 E DSTR              ACKNOWLEDGED - #16->#15 P1K03 - RESOURCE_DELETION DELETE for partition 2
3 E RESOURCE_DELETION DELETING     - #05->#04 P1K03 - ResourceDeletionRecord {"tenantId":"<default>","resourceKey":2251799813685250}
3 E PROC              DELETING     - #06->#04 P3K01 - process.xml -> "id-e756..45a8a2c" (version:1)
3 E PROC              DELETED      - #07->#04 P3K02 - process.xml -> "id-e756..45a8a2c" (version:1)
3 E RESOURCE_DELETION DELETED      - #08->#04 P1K03 - ResourceDeletionRecord {"tenantId":"<default>","resourceKey":2251799813685250}
1 C DSTR              ACKNOWLEDGE  - #17-> -1 P1K03 - RESOURCE_DELETION DELETE for partition 3
1 E DSTR              ACKNOWLEDGED - #18->#17 P1K03 - RESOURCE_DELETION DELETE for partition 3
1 E DSTR              FINISHED     - #19->#17 P1K03 - RESOURCE_DELETION DELETE on partition 1
2 C DPLY              CREATE       - #09-> -1 P1K01 - process.xml
3 C DPLY              CREATE       - #09-> -1 P1K01 - process.xml
2 E PROC              CREATED      - #10->#09 P1K02 - process.xml -> "id-e756..45a8a2c" (version:1)
2 E DPLY              CREATED      - #11->#09 P1K01 - process.xml
3 E PROC              CREATED      - #10->#09 P1K02 - process.xml -> "id-e756..45a8a2c" (version:1)
3 E DPLY              CREATED      - #11->#09 P1K01 - process.xml
```

</p>
</details> 

**To Reproduce**

<details><summary>Reproducing test case</summary>
<p>

```java
@Test
public void shouldDeleteIdempotently() {
  // given the deployment cannot be acknowledged
  engine.interceptInterPartitionCommands(
      (receiverPartitionId, valueType, intent, recordKey, command) ->
          !(receiverPartitionId == 1
              && intent == CommandDistributionIntent.ACKNOWLEDGE
              && ((CommandDistributionRecordValue) command).getValueType()
                  == ValueType.DEPLOYMENT));

  final var processId = Strings.newRandomValidBpmnId();
  final long resourceKey =
      engine
          .deployment()
          .withXmlResource(Bpmn.createExecutableProcess(processId).startEvent().endEvent().done())
          .deploy()
          .getValue()
          .getProcessesMetadata()
          .getFirst()
          .getProcessDefinitionKey();

  // when
  engine.resourceDeletion().withResourceKey(resourceKey).delete();
  RecordingExporter.commandDistributionRecords(CommandDistributionIntent.FINISHED)
      .withDistributionIntent(ResourceDeletionIntent.DELETE)
      .await();

  // push time forward to trigger the retry
  // seems that the command redistributor needs several pushes to trigger the retry
  engine.increaseTime(Duration.ofMinutes(1));
  engine.increaseTime(Duration.ofMinutes(1));
  engine.increaseTime(Duration.ofMinutes(1));
  engine.increaseTime(Duration.ofMinutes(1));
  engine.increaseTime(Duration.ofMinutes(1));

  // retried deployments should be rejected after deletion
  RecordingExporter.deploymentRecords(DeploymentIntent.CREATE)
      .withPartitionId(2)
      .onlyCommandRejections()
      .await();
  RecordingExporter.deploymentRecords(DeploymentIntent.CREATE)
      .withPartitionId(3)
      .onlyCommandRejections()
      .await();

  // todo: should not be able to create instance of process on any partition
}
```

</p>
</details> 

**Expected behavior**

A deleted resource should remain deleted.

Either:
- retried deployments must be rejected after that resource has been deleted previously
- or once a resource is requested to be deleted, stop (re)distributing the deployment

**Comment by korthout** [comment]
I believe this is a high severity (it's unexpected and can lead to problems in production), but there is a low likelihood of this actually occurring (requires very specific conditions).

There is no workaround: the user can't delete the resource a second time (rejected as not found by partition 1).

**Comment by korthout** [comment]
I've chosen for a bug title that conveys user impact, but this is not restricted to process deployments only. Other resources, like decisions and forms are also affected

**Comment by mustafadagher** [comment]
ZPA triage:
- Might (or might not) be resolved by #21447 
- When  #21447 is introduced we shall come back and change how we use the command distribution accordingly
